File: Python/ByTier/B/1938_B_Attraction_Score.py
# Problem: CF 1938 B - Attraction Score
# https://codeforces.com/contest/1938/problem/B

"""
Algorithm: Maximum Attraction Score
Approach: 
This solution uses a greedy approach combined with dynamic programming techniques to find the maximum attraction score.
The algorithm processes nodes in a specific order (based on degree), and for each node, it examines:
1. All pairs of neighbors connected to the current node (for 2-cliques)
2. All triplets of neighbors (for 3-cliques)
3. All quadruplets of neighbors (for 4-cliques)

Key insight: 
- Nodes with degree <= 5 are processed in a queue for optimization
- For each processed node, the algorithm evaluates potential maximum scores through different combinations of its neighbors
- A penalty term is computed for each subset with disconnected pairs of cities
- Memoization is used to store and reuse computed scores for combinations of cities (SZ2 for pairs, SZ3 for triplets)

Time Complexity: O(N * D^4) where N is number of nodes and D is max degree (bounded by 5 in the queue processing)
Space Complexity: O(N * D^2) for storing adjacency lists and memoization tables

Note: The queue-based processing and degree constraint are key optimizations that make this feasible for the given constraints.
"""

import sys
from collections import deque
from itertools import combinations

input = sys.stdin.readline
F = lambda: [*map(int, input().split())]
penalty = 10**6


def rmv(x):
    # Remove node x from adjacency lists of its neighbors
    for y in ADJ:
        G[y].pop(x)
        # Add neighbor to queue if degree drops to 5 or less
        if VIS[y] == 0 and len(G[y]) <= 5:
            Q.append(y)
            VIS[y] = 1


N, M = F()
# Read coordinates (but not used in solution - only for input parsing)
for _ in range(N):
    input()
# Initialize adjacency list
G = [{} for _ in range(N)]
# Read edges
for _ in range(M):
    a, b, v = F()
    a -= 1  # Convert to 0-indexed
    b -= 1
    G[a][b] = v  # Store edge weight
    G[b][a] = v  # Edge is undirected

# Queue for processing nodes with small degree
Q = deque()
SZ2 = {}  # Memoization for pairs of cities
SZ3 = {}  # Memoization for triplets of cities

VIS = [0] * N
# Add all nodes with degree <= 5 to queue for processing
for i in range(N):
    if len(G[i]) <= 5:
        Q.append(i)
        VIS[i] = 1
ans = 0  # Maximum attraction score found
while Q:
    x = Q.popleft()  # Get next node to process
    l = len(G[x])  # Degree of current node
    if l == 0:  # Skip isolated nodes
        continue
    ADJ = []
    # Build list of adjacent nodes
    for y in G[x]:
        ADJ.append(y)
        # Consider single highway as potential score
        ans = max(ans, G[x][y])

    if l == 1:  # If node is connected to only 1 other, remove it
        rmv(x)
        continue

    # Process pairs of neighbors of the current node
    for a, b in combinations(ADJ, 2):
        if b not in G[a]:  # Skip if no direct connection between neighbors
            continue
        # Compute score for this triangle
        s = G[x][a] + G[x][b] + G[a][b]
        ans = max(ans, s)

        # Update memoization tables for pairs
        K = sorted([a, b, x])
        for i in range(2):
            for j in range(i + 1, 3):
                aa, bb = K[i], K[j]
                if (aa, bb) in SZ2:
                    ans = max(ans, SZ2[(aa, bb)] + s)
                    SZ2[(aa, bb)] = max(SZ2[(aa, bb)], s - G[aa][bb] - penalty)
                else:
                    SZ2[(aa, bb)] = s - G[aa][bb] - penalty
    if l == 2:  # If node has degree 2, remove it
        rmv(x)
        continue

    # Process triplets of neighbors of the current node
    for a, b, c in combinations(ADJ, 3):
        cnt = 0  # Count of disconnected pairs
        s = G[x][a] + G[x][b] + G[x][c]
        # Update score with edges connecting x to neighbors
        if b in G[a]:
            s += G[a][b]
        else:
            cnt += 1
            s -= penalty
        if c in G[a]:
            s += G[a][c]
        else:
            cnt += 1
            s -= penalty
        if c in G[b]:
            s += G[b][c]
        else:
            cnt += 1
            s -= penalty
        ans = max(ans, s)
        if cnt > 0:  # Skip if more than 1 disconnected pair (too expensive)
            continue

        # Update memoization for triplets
        K = sorted([a, b, c, x])
        aa, bb, cc, dd = K[0], K[1], K[2], K[3]
        if (aa, bb, cc) in SZ3:
            ans = max(ans, SZ3[(aa, bb, cc)] + s)
            SZ3[(aa, bb, cc)] = max(
                SZ3[(aa, bb, cc)], G[dd][aa] + G[dd][bb] + G[dd][cc] - penalty
            )
        else:
            SZ3[(aa, bb, cc)] = G[dd][aa] + G[dd][bb] + G[dd][cc] - penalty

        # Try all permutations of the triplet to ensure all combinations are covered
        aa, bb, cc, dd = K[0], K[1], K[3], K[2]
        if (aa, bb, cc) in SZ3:
            ans = max(ans, SZ3[(aa, bb, cc)] + s)
            SZ3[(aa, bb, cc)] = max(
                SZ3[(aa, bb, cc)], G[dd][aa] + G[dd][bb] + G[dd][cc] - penalty
            )
        else:
            SZ3[(aa, bb, cc)] = G[dd][aa] + G[dd][bb] + G[dd][cc] - penalty

        aa, bb, cc, dd = K[0], K[2], K[3], K[1]
        if (aa, bb, cc) in SZ3:
            ans = max(ans, SZ3[(aa, bb, cc)] + s)
            SZ3[(aa, bb, cc)] = max(
                SZ3[(aa, bb, cc)], G[dd][aa] + G[dd][bb] + G[dd][cc] - penalty
            )
        else:
            SZ3[(aa, bb, cc)] = G[dd][aa] + G[dd][bb] + G[dd][cc] - penalty

        aa, bb, cc, dd = K[1], K[2], K[3], K[0]
        if (aa, bb, cc) in SZ3:
            ans = max(ans, SZ3[(aa, bb, cc)] + s)
            SZ3[(aa, bb, cc)] = max(
                SZ3[(aa, bb, cc)], G[dd][aa] + G[dd][bb] + G[dd][cc] - penalty
            )
        else:
            SZ3[(aa, bb, cc)] = G[dd][aa] + G[dd][bb] + G[dd][cc] - penalty
    
    if l == 3:  # If node has degree 3, remove it
        rmv(x)
        continue

    # Process quadruplets of neighbors of the current node
    for a, b, c, d in combinations(ADJ, 4):
        cnt = 0
        s = G[x][a] + G[x][b] + G[x][c] + G[x][d]
        # Update score with edges connecting x to neighbors
        if b in G[a]:
            s += G[a][b]
        else:
            cnt += 1
            s -= penalty
        if c in G[a]:
            s += G[a][c]
        else:
            cnt += 1
            s -= penalty
        if d in G[a]:
            s += G[a][d]
        else:
            cnt += 1
            s -= penalty
        if c in G[b]:
            s += G[b][c]
        else:
            cnt += 1
            s -= penalty
        if d in G[b]:
            s += G[b][d]
        else:
            cnt += 1
            s -= penalty
        if d in G[c]:
            s += G[c][d]
        else:
            cnt += 1
            s -= penalty
        if cnt >= 2:  # Too many disconnected pairs, skip
            continue
        ans = max(ans, s)

    rmv(x)  # Remove processed node

print(ans)  # Output maximum attraction score


# https://github.com/VaHiX/CodeForces/

File: CPP/ByTier/B/1938_B_Attraction_Score.cpp
// Problem: CF 1938 B - Attraction Score
// https://codeforces.com/contest/1938/problem/B

/*
 * Purpose: Find the maximum attraction score among all possible non-empty subsets of cities
 *          in a graph with weighted edges, where the score includes attraction points from 
 *          connected cities and penalizes disconnected pairs.
 *
 * Algorithm: 
 * - The code uses a combination of geometric analysis (cross products) and graph traversal
 *   to evaluate potential subsets and compute optimal attraction scores.
 * - It processes edges in a specific order, maintaining maximum attraction values for each node.
 * - Cross products are used to determine orientation of points and identify potential triangle
 *   configurations that contribute to the score.
 * - The solution considers both direct edge weights and combinations involving up to three edges.
 * 
 * Time Complexity: O(m^2) in worst case due to nested iterations over edges,
 *                  but optimized in practice with careful data structures.
 * Space Complexity: O(n + m) for storing graph data, coordinates, and auxiliary arrays.
 */

#include <algorithm>
#include <iostream>
#include <map>
#include <utility>

#pragma GCC optimize("O2")
#define ll long long
#define pb emplace_back
#define mp make_pair
#define orz 1000000007
#define fi first
#define se second
using namespace std;
int n, m, x[300005], y[300005], z[300005];
int ans, d[100005], l[100005], t[300005], I[300005], w[300005], u[300005];
int mx[300005][2];
pair<ll, int> fmx[300005], fmn[300005];
int fx[300005], fn[300005];
pair<int, int> p[100005];
map<int, int> M[100005];
void U(int x, int y) {
  if (y > mx[x][0])
    mx[x][1] = mx[x][0], mx[x][0] = y;
  else if (y > mx[x][1])
    mx[x][1] = y;
}
void u2(int i, int z, int _) {
  int X = x[i], Y = y[i];
  pair<int, int> p1, p2;
  p1 = mp(p[Y].fi - p[X].fi, p[Y].se - p[X].se);
  p2 = mp(p[z].fi - p[Y].fi, p[z].se - p[Y].se);
  ll o = p1.fi * 1ll * p2.se - p1.se * 1ll * p2.fi;
  if (o > fmx[i].fi)
    fmx[i] = mp(o, z), fx[i] = _;
  if (o < fmn[i].fi)
    fmn[i] = mp(o, z), fn[i] = _;
}
ll dis(int x, int y) {
  ll X = p[x].fi - p[y].fi, Y = p[x].se - p[y].se;
  return X * 1ll * X + Y * 1ll * Y;
}
int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);
  cout.tie(0);
  cin >> n >> m;
  for (int i = 1; i <= n; ++i)
    cin >> p[i].fi >> p[i].se;
  for (int i = 1; i <= m; ++i)
    cin >> x[i] >> y[i] >> z[i], ans = max(ans, z[i]), ++d[x[i]], ++d[y[i]],
                                 M[x[i]][y[i]] = i, M[y[i]][x[i]] = i;
  for (int i = 1; i <= m; ++i) {
    if (mp(d[x[i]], x[i]) > mp(d[y[i]], y[i]))
      swap(x[i], y[i]);
    ++l[x[i]];
  }
  for (int i = 1; i <= m; ++i)
    mx[i][0] = mx[i][1] = -1;
  for (int i = 1; i <= n; ++i)
    l[i + 1] += l[i];
  for (int i = 1; i <= m; ++i) {
    t[l[x[i]]] = y[i];
    I[l[x[i]]] = i;
    w[l[x[i]]--] = z[i];
  }
  for (int i = 1; i <= n; ++i) {
    for (int j = l[i + 1]; j != l[i]; --j)
      u[t[j]] = I[j];
    for (int j = l[i + 1]; j != l[i]; --j) {
      int o = t[j];
      for (int k = l[o + 1]; k != l[o]; --k) {
        if (u[t[k]]) {
          ans = max(ans, w[j] + w[k] + z[u[t[k]]]);
          U(I[j], w[k] + z[u[t[k]]]);
          U(I[k], w[j] + z[u[t[k]]]);
          U(u[t[k]], w[j] + w[k]);
          u2(I[j], t[k], w[k] + z[u[t[k]]]);
          u2(I[k], i, w[j] + z[u[t[k]]]);
          u2(u[t[k]], o, w[j] + w[k]);
        }
      }
    }
    for (int j = l[i + 1]; j != l[i]; --j)
      u[t[j]] = 0;
  }
  for (int i = 1; i <= m; ++i)
    if (mx[i][1] != -1)
      ans = max(ans, z[i] + mx[i][0] + mx[i][1] - 1000000);
  for (int i = 1; i <= m; ++i)
    if (fmx[i].se && fmn[i].se) {
      int o = M[fmx[i].se][fmn[i].se];
      if (o) {
        int bas = z[i] + fx[i] + fn[i] + z[o];
        int X, Y;
        if (dis(x[i], fmn[i].se) > dis(y[i], fmn[i].se))
          X = x[i], Y = y[i];
        else
          X = y[i], Y = x[i];
        ans = max(ans, bas);
        bas -= 1000000;
        int ix = M[X][fmx[i].se];
        int in = M[X][fmn[i].se];
        if (fmx[ix].se == fmx[in].se && fmx[ix].se != Y && fmx[ix].se != 0) {
          int Z = fmx[ix].se;
          if (M[X].count(Z))
            ans = max(ans, bas + fx[ix] + fx[in] - z[M[X][Z]]);
        }
        if (fmx[ix].se == fmn[in].se && fmx[ix].se != Y && fmx[ix].se != 0) {
          int Z = fmx[ix].se;
          if (M[X].count(Z))
            ans = max(ans, bas + fx[ix] + fn[in] - z[M[X][Z]]);
        }
        if (fmn[ix].se == fmx[in].se && fmn[ix].se != Y && fmn[ix].se != 0) {
          int Z = fmn[ix].se;
          if (M[X].count(Z))
            ans = max(ans, bas + fn[ix] + fx[in] - z[M[X][Z]]);
        }
        if (fmn[ix].se == fmn[in].se && fmn[ix].se != Y && fmn[ix].se != 0) {
          int Z = fmn[ix].se;
          if (M[X].count(Z))
            ans = max(ans, bas + fn[ix] + fn[in] - z[M[X][Z]]);
        }
      }
    }
  cout << ans << '\n';
  return 0;
}


// https://github.com/VaHiX/CodeForces/